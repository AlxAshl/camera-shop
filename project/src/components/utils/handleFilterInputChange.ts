import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';

export const handleFilterInputChange = (filterName: string, filter: string[], action: ActionCreatorWithNonInferrablePayload) => {
  const mutableFilter = Object.assign([], filter);
  if(filter.includes(filterName)) {
    const updatedFilter = mutableFilter.filter((entry) => entry !== filterName);
    store.dispatch(action(updatedFilter));
  }
  else {
    mutableFilter.push(filterName);
    store.dispatch(action(mutableFilter));
  }
};
