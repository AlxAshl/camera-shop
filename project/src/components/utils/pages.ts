import { PAGE_LIMIT } from '../../const';

function getPagesCount(cameraCount: number): number {
  return Math.ceil(cameraCount / PAGE_LIMIT);
}

export default getPagesCount;
