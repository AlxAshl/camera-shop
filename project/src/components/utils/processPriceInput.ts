import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';
import { ProductType } from '../../types/product';

export const setMinInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMinPriceInput;
    const newMin = productsArray.find((product) => product.price <= Number(input));
    newMin?.price === undefined
      ? newMinPriceInput = productsArray[productsArray.length - 1].price
      : newMinPriceInput = newMin?.price;
    store.dispatch(action([String(newMinPriceInput)]));
  }
};

export const setMaxInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMaxPriceInput;
    const newMax = productsArray.find((product) => product.price >= Number(input));
    newMax?.price === undefined
      ? newMaxPriceInput = productsArray[productsArray.length - 1].price
      : newMaxPriceInput = newMax?.price;
    store.dispatch(action([String(newMaxPriceInput)]));
  }
};
