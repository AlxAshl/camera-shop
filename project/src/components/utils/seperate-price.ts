export function seperatePrice(priceValue: number) {
  let returnedPrice = String(priceValue);
  if(returnedPrice.length > 6) {
    returnedPrice = `${returnedPrice.slice(0, 1)} ${returnedPrice.slice(1, 4)} ${returnedPrice.slice(-3)}`;
  }
  if (returnedPrice.length > 3) {
    returnedPrice = `${returnedPrice.slice(-0, -3)} ${returnedPrice.slice(-3)}`;
  }
  return returnedPrice;
}
