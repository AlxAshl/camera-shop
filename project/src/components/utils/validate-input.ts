export function validateInput(target: (EventTarget & HTMLInputElement) | (EventTarget & HTMLTextAreaElement)) {

  if (target.parentElement?.parentElement && target.name === 'rating' && (Number(target.value) > 0)) {
    if (!target.parentElement.parentElement.parentElement?.className.includes('is-valid')) {
      target.parentElement.parentElement.parentElement?.classList.add('is-valid');
      target.parentElement.parentElement.parentElement?.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name !== 'review' && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length > 0) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string'){
    if(target.value.length === 0) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length >= 5) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name === 'review' && target.value.length < 5) {
    if (!target.parentElement.parentElement.className.includes('is-invalid')) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

}
