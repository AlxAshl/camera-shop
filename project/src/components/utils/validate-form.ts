import { reviewDataType } from '../../types/review';
import { validateInput } from './validate-input';


export function validateForm(data: reviewDataType) {
  let validationStatus = true;
  Object.entries(data).forEach(([key, value]) =>{
    if (key !== 'review') {
      if(value === '') {
        const failedInput = document.querySelector(`input[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
        validationStatus = false;
      }
    }
    if(key === 'review') {
      if(String(value).length < 5) {
        const failedInput = document.querySelector(`textarea[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
      }
    }
    if(key === 'rating') {
      if(value < 1) {
        const failedInput = document.querySelector(`input[name=${key}]`);
        if(failedInput && failedInput.parentElement?.parentElement){
          failedInput.parentElement.parentElement.parentElement?.classList.add('is-invalid');
        }
        validationStatus = false;
      }
    }
  });
  return validationStatus;
}

