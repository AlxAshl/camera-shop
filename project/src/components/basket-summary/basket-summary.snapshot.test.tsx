import { renderWithProviders } from '../../test/utils/render-with-redux';
import BasketSummary from './basket-summary';

test('Render BasketSummary', () => {
  const { asFragment } = renderWithProviders(<BasketSummary/>);
  expect(asFragment()).toMatchSnapshot();
});
