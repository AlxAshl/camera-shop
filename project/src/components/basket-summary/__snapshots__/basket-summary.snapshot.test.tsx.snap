// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render BasketSummary 1`] = `
<DocumentFragment>
  <div
    class="basket__summary"
  >
    <div
      class="basket__promo"
    >
      <p
        class="title title--h4"
      >
        Если у вас есть промокод на скидку, примените его в этом поле
      </p>
      <div
        class="basket-form"
      >
        <form
          action="#"
        >
          <div
            class="custom-input"
          >
            <label>
              <span
                class="custom-input__label"
              >
                Промокод
              </span>
              <input
                name="promo"
                placeholder="Введите промокод"
                type="text"
              />
            </label>
            <p
              class="custom-input__error"
            >
              Промокод неверный
            </p>
            <p
              class="custom-input__success"
            >
              Промокод принят!
            </p>
          </div>
          <button
            class="btn"
            type="submit"
          >
            Применить
          </button>
        </form>
      </div>
    </div>
    <div
      class="basket__summary-order"
    >
      <p
        class="basket__summary-item"
      >
        <span
          class="basket__summary-text"
        >
          Всего:
        </span>
        <span
          class="basket__summary-value"
        >
          111 390 ₽
        </span>
      </p>
      <p
        class="basket__summary-item"
      >
        <span
          class="basket__summary-text"
        >
          Скидка:
        </span>
        <span
          class="basket__summary-value basket__summary-value--bonus"
        >
          0 ₽
        </span>
      </p>
      <p
        class="basket__summary-item"
      >
        <span
          class="basket__summary-text basket__summary-text--total"
        >
          К оплате:
        </span>
        <span
          class="basket__summary-value basket__summary-value--total"
        >
          111 390 ₽
        </span>
      </p>
      <button
        class="btn btn--purple"
        type="submit"
      >
        Оформить заказ
      </button>
    </div>
  </div>
</DocumentFragment>
`;
